craft_application/commands/init.py

```diff
@@ -101,8 +102,19 @@ def profiles(self) -> list[str]:
 
     def run(self, parsed_args: argparse.Namespace) -> None:
         """Run the command."""
-        project_dir = self._get_project_dir(parsed_args)
+        # If the user provided a "name" and it's not valid, the command fails.
+        if parsed_args.name is not None:
+            self._services.init.validate_project_name(parsed_args.name)
+
+        # However, if the name comes from the directory, we don't fail and
+        # instead fallback to "my-project".
         project_name = self._get_name(parsed_args)
+        try:
+            self._services.init.validate_project_name(project_name)
+        except errors.InitError:
+            project_name = "my-project"
```

Comment:
What about moving this logic to `self._services.init.validate_project_name(project_name)` and if you specify `fallback=` it will return it instead of erroring out ?